utils/yoke/src/cart.rs

```diff
@@ -0,0 +1,120 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Yoke, Yokeable};
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// A [`Cart`] is a type that is acceptable as backing storage in a [`Yoke`].
+///
+/// The essential invariant that must be maintained by implementors is that `Self::Inner` references
+/// obtained from this type via `Self::get_inner()` must continue to be valid for the
+/// complete lifetime of this type (i.e. until destructors run) provided that this type
+/// is never accessed via `&mut` references.
+///
+/// For example, `Rc<Vec<u8>>` and `Rc<[u8]>` are valid [`Cart`]s, however `Rc<RefCell<Vec<u8>>>` is
+/// not, because in the latter type it is possible to use interior mutation to change the data.
+///
+/// In general, this means "no interior mutability", though interior mutability can be fine
+/// if the interior mutability is to something that does not affect references. For example,
+/// `Rc<[u8]>` does have interior mutability in the refcount, but you cannot get references
+/// to it, so it's fine. On the other hand, `Weak<Box<[u8]>>` cannot be a valid [`Cart`] because
+/// it is possible for the backing buffer to be cleaned up without warning.
```

Comment:
Right, and then this means that as a corollary, `Rc<Vec<RefCell<u8>>>` is ok but the transposition `Rc<RefCell<Vec<u8>>` is not, because you might cause the Vec's pointee address to change?